robot_calibration/src/util/action_server.cpp

```diff
@@ -0,0 +1,155 @@
+      {
+        RCLCPP_WARN(logger, "Unable to move to desired state for sample %u.", pose_idx);
+        continue;
+      }
+
+      // Make sure sensor data is up to date after settling
+      rclcpp::sleep_for(std::chrono::milliseconds(2000));
+
+      // Get pose of the features
+      if (!capture_manager.captureFeatures(robot_poses[pose_idx].features, msg))
+      {
+        RCLCPP_WARN(logger, "Failed to capture sample %u.", pose_idx);
+        continue;
+      }
+
+      RCLCPP_INFO(logger, "Captured pose %u of %lu", pose_idx + 1, robot_poses.size());
+
+      // Add to samples
+      data.push_back(msg);
+    }
+
+    RCLCPP_INFO(logger, "Done capturing samples");
+  }
+
+  // Create instance of optimizer
+  robot_calibration::OptimizationParams params;
+  robot_calibration::Optimizer opt(description_msg.data);
+
+  // Load calibration steps
+  std::vector<std::string> calibration_steps =
+      node_->declare_parameter<std::vector<std::string>>("calibration_steps", std::vector<std::string>());
+  if (calibration_steps.empty())
+  {
+    RCLCPP_FATAL(logger, "Parameter calibration_steps is not defined");
+  }
+
+  // Run calibration steps
+  for (auto step : calibration_steps)
+  {
+    params.LoadFromROS(node_, step);
+
+    // If we receive initial guess from action goal, override 
+    std::string initial_guess_param = step + ".free_frames_initial_values";
+    RCLCPP_INFO(logger, "Overriding initial values for: %s", initial_guess_param.c_str());
+
+    for (unsigned i=0; i < params.free_frames.size(); i++) {
+      for (const auto& estimated_pose : goal->estimated_poses) {
+        if (estimated_pose.child_frame_id == params.free_params.at(i)) {
+          params.free_frames_initial_values.at(i).x = estimated_pose.transform.translation.x;
+          params.free_frames_initial_values.at(i).y = estimated_pose.transform.translation.y;
+          params.free_frames_initial_values.at(i).z = estimated_pose.transform.translation.z;
+
+          tf2::Quaternion q(estimated_pose.transform.rotation.x,
+                            estimated_pose.transform.rotation.y,
+                            estimated_pose.transform.rotation.z,
+                            estimated_pose.transform.rotation.w);
+          tf2::Matrix3x3 m(q);
+          double roll, pitch, yaw;
+          m.getRPY(roll, pitch, yaw);
+
+          params.free_frames_initial_values[i].roll = roll;
+          params.free_frames_initial_values[i].pitch = pitch;
+          params.free_frames_initial_values[i].yaw = yaw;
+        }
+      }
+    }
+
+    opt.optimize(params, data, logger, true);
+
+    std::cout << "Parameter Offsets:" << std::endl;
+    std::cout << opt.getOffsets()->getOffsetYAML() << std::endl;
+  }
+
+  // Write outputs
+  // robot_calibration::exportResults(opt, description_msg.data, data);
+  auto result = std::make_shared<CalibratePose::Result>();
+
+  RCLCPP_INFO(logger, "Done calibrating");
```

Comment:
Can we add the result with a TODO that the pose's need to be filled in
```
// If we make it here, running the calibration succeeded
auto calibrate_action_result = std::make_shared<CalibratePose::Result>();
// calibrate_action_result->calibrated_poses.push_back(frame);  // TODO fill in the answer
has_active_action_goal_ = false;
goal_handle->succeed(action_result);
```

Reply:
Added the result from the cal process now